### ISSUE TITLE
[Extractor] Deploying multiple extracted apis: PessimisticConcurrencyConflict,  Operation on the API is in progress

### ISSUE THREAD

**Issue by bremnes** [Azure/azure-api-management-devops-resource-kit#340]
### Context
We have organized our apim arm template project according to the proposed structure
![image](https://user-images.githubusercontent.com/3669798/73432160-7222b080-4342-11ea-81fa-ba68517f3e2d.png)

We are using the extractor, and then have an additional tool slightly manipulating and inserting the extracted templates into our arm template solution. We have this folder structure in our arm template project (*.deployproj):

    .
    ├── master.template.json
    ├── master.parameters.json
    ├── named-values.template.json          # linked to from master
    ├── tags.template.json                  # linked to from master
    ├── Api/
        ├── apis.template.json              # linked to from master, sub-master links to all apis
        ├── petstore/
            ├── petstore.template.json
            ├── petstore-apiPolicy.xml
            ├── get-pet-operationPolicy.xml
        ├── another-api/
            ├── another-api.template.json
            ├── another-api-apiPolicy.xml
        └── ...
    ├── Product/
        ├── products.template.json          # linked to from master, sub-master links to all products
        ├── unlimited.json
        ├── unlimited-productPolicy.xml
        ├── starter.json
        ├── petstore.json
        ├── petstore-productPolicy.xml
        └── ...
    └── ...

In the master template we have dependencies between the "sub-master"s like this (excluding a lot of properties and parameters):
```json
{
    "name": "apisTemplate",
    "type": "Microsoft.Resources/deployments",
    "dependsOn": [
        "namedValuesTemplate",
        "apiVersionSetListTemplate",
        "productsTemplate"
    ]
}, {
    "name": "namedValuesTemplate",
    "type": "Microsoft.Resources/deployments",
    "dependsOn": []
}, {
    "name": "productsTemplate",
    "type": "Microsoft.Resources/deployments",
    "dependsOn": []
}, {
    "name": "tagsTemplate",
    "type": "Microsoft.Resources/deployments",
    "dependsOn": []
}
```

When using the extractor it adds the product<->api binding in the bottom of the api templates, and that resource looks like this:
```json
{
    "name": "[concat(parameters('ApimServiceName'), '/unlimited/petstore')]",
    "type": "Microsoft.ApiManagement/service/products/apis",
    "apiVersion": "2019-01-01",
    "dependsOn": [
        "[resourceId('Microsoft.ApiManagement/service/apis', parameters('ApimServiceName'), 'petstore')]"
    ]
}
```

### Problem
We currently have ~40 apis and ~30 products. When deploying this to the azure resource manager we have gotten more and more errors like this. It is not consistent on which template it fails, or even how many templates fails. Deployment fails 60-70% of the time, and we just have to redeploy until it works.
```
"properties": {
    "statusCode": "Conflict",
    "statusMessage": "{\"error\":{\"code\":\"PessimisticConcurrencyConflict\",\"message\":\"Operation on the API is in progress\",\"details\":null}}"
}
```
In the activity log it either fails with this message on either "Create API or Update API" or "Add API to product".

### Azure support
An issue with support was raised, #120012721000749 for those of you with access, and we got the feedback that there was a lock/race situation going on. When asking further it seems like the resource type `Microsoft.ApiManagement/service/products/apis` places a lock on _both_ the api and product.

### Solution?
We tried to flip the situation and did a quick POC where we moved the product<->api binding to the product file (and also reversed the master `dependsOn` between apisTemplate and productsTemplate). We did this because it seemed like the resource `Microsoft.ApiManagement/service/products/apis` belongs to product. This _seems_ to work 100% of the time. However, I'm unsure if this is a long term solution since support said that the lock is on both api and product. If that's true, it means that we are just "lucky". Because the products are deployed in parallel and several of them has bindings to the same api.

### Help
Any thoughts..? Asking here since the extractor currently adds the api<->product binding in the api template file.

**Comment by RupengLiu** [comment]
@bremnes thanks for reporting the issue!
The deployment process failed when multiple apis want to be added to the same product by deploying the resources 
```
{
    "name": "[concat(parameters('ApimServiceName'), '/unlimited/petstore')]",
    "type": "Microsoft.ApiManagement/service/products/apis",
    "apiVersion": "2019-01-01",
    "dependsOn": [
        "[resourceId('Microsoft.ApiManagement/service/apis', parameters('ApimServiceName'), 'petstore')]"
    ]
}
``` 
So the current implementation of our resource kit will cause this deployment bug.
As for the solution you mentioned above, it will work 100% when you are adding new API, products, or the product already exist on the destination apim instance. 
The most ideal way to deal with this situation is to have separate product-api relationship template, which depends on both apiTemplate / productTemplate

**Comment by bremnes** [comment]
Thanks for the quick reply @RupengLiu!

> As for the solution you mentioned above, it will work 100% when you are adding new API, products, or the product already exist on the destination apim instance.

I'm not sure about this. In our current arm template structure, where it fails 60-70% of the time, it still can fail with the lock exception even though all resources already exists. The only way we get a 100% success ratio is if we move the api-product binding to the product template (both for apim instances where all resources exists, and for instances which are blank). But since you guys are saying that `Microsoft.ApiManagement/service/products/apis` places a lock on both api and product, I'm not confident in that solution.

My impression is that the dependsOn only can be used towards resources in the same template, in addition it doesn't guarantee order, just that the dependecies will be deployed successfully first.

Let's say that we have 5 products and 5 apis and, for the scenario's sake, all the products are connect to all the apis. Can you "paint a picture" of how this can be organized? The way I see it we still will get a parellel deployment of those product-api resources and having the risk of a lock collision.

If we need to execute the api-product binding resources in sequence to avoid locks/collisions, the only way I see possible is something like this. But it seems extremely messy:
```
{
    "name": "apisTemplate",
    "type": "Microsoft.Resources/deployments",
    "dependsOn": [
        "namedValuesTemplate",
        "apiVersionSetListTemplate"
    ]
}, {
    "name": "namedValuesTemplate",
    "type": "Microsoft.Resources/deployments",
    "dependsOn": []
}, {
    "name": "productsTemplate",
    "type": "Microsoft.Resources/deployments",
    "dependsOn": [
        "namedValuesTemplate"
    ]
}, {
    "name": "tagsTemplate",
    "type": "Microsoft.Resources/deployments",
    "dependsOn": []
}, {
    "name": "apiAproductAbinding",
    "type": "Microsoft.Resources/deployments",
    "dependsOn": [
        "productsTemplate",
        "apisTemplate"
    ]
}, {
    "name": "apiAproductBbinding",
    "type": "Microsoft.Resources/deployments",
    "dependsOn": [
        "apiAproductAbinding"
    ]
}, {
    "name": "apiAproductCbinding",
    "type": "Microsoft.Resources/deployments",
    "dependsOn": [
        "apiAproductBbinding"
    ]
}
...
```

**Comment by RupengLiu** [comment]
Hi @bremnes , I checked with the APIM backend team, the deployment failure is caused by adding the same api to different products at the same time. When adding apis to product, will in fact create new revision of the Api, if it's running concurrently, the write to the same api will cause a race, and currently this situation will not change. So you are right about creating dependencies for a series of api-product relation templates, but what's different is that we should create the dependencies within each apiTemplate, since we only need to make sure that adding api to multiple products happens sequentially.

**Comment by bremnes** [comment]
Hi again @RupengLiu,

I revisited this again and tried your suggest approach - I ran a script on all our api templates where I created a dependsOn "chain" so that the product<->api bindings were deployed sequentially for each api.

Here's the result of deploying 15 times:
* 4 failure with `PessimisticConcurrencyConflict` `Operation on the API is in progress`
* 1 timed out/failed. Ran for 45 minutes before I killed the deployment (it usually runs in 2-5 minutes when successfully).
* 1 failed with `PessimisticConcurrencyConflict` `Operation on repositories is in progress` (never seen this before)
* 9 ran OK

So it seems like it's some kind of racing condition on both the apis and the products - meaning we can't deploy api-product bindings in parallell at all.


I tested one final thing where I introduced a `productApiBindingsTemplate` linked to from master which has a dependency to `productsTemplate` and `apisTemplate`. I moved all api-product bindings inside that file and then made them run sequentially by chaining the `dependsOn` attribute. This works and deploys successfully 14 of 15 times, but to be honest I'm not a big fan of this solution..

On to more exploring. Perhaps I'll try deploying the apis sequentially or something.

**Comment by awnugent** [comment]
I'm running into this same issue. 
@bremnes, did you have any luck with deploying apis sequentially? I'm considering the same strategy.

**Comment by bremnes** [comment]
Hi @awnugent,

I tried several approaches, but it all comes down to this locking situation where the possibility of failure increases exponentially the more apis you try to deploy in parallel. We talked with Microsoft as well, and they confirmed our thoughts and advised to use `dependsOn` as well. So what I ended up with was to deploy sequentially in our apis.template.json. It takes longer, but at least it works every time. (Sidenote: we will probably split the apis into smaller deployable units.)

So basically something like this. As mentioned earlier we've created a wrapper around the extractor which does additional things like organizing the extracted templates into our arm template project, and now also chain the `dependsOn` attribute:
```json
...
{
	"apiVersion": "2017-05-10",
	"name": "petstoreTemplate",
	"type": "Microsoft.Resources/deployments",
	"properties": {
		"mode": "Incremental",
		"templateLink": {
			"uri": "[concat(parameters('_artifactsLocation'), '/Api/petstore/petstore.template.json', parameters('_artifactsLocationSasToken'))]",
			"contentVersion": "1.0.0.0"
		},
		"parameters": {
			"url, artifacts location, sas token and apim instance name"
		}
	},
	"dependsOn": [
	]
}, {
	"apiVersion": "2017-05-10",
	"name": "echoTemplate",
	"type": "Microsoft.Resources/deployments",
	"properties": {
		"mode": "Incremental",
		"templateLink": {
			"uri": "[concat(parameters('_artifactsLocation'), '/Api/echo/echo.template.json', parameters('_artifactsLocationSasToken'))]",
			"contentVersion": "1.0.0.0"
		},
		"parameters": {
			"url, artifacts location, sas token and apim instance name"
		}
	},
	"dependsOn": [
		"petstoreTemplate"
	]
}, {
	"apiVersion": "2017-05-10",
	"name": "wikiTemplate",
	"type": "Microsoft.Resources/deployments",
	"properties": {
		"mode": "Incremental",
		"templateLink": {
			"uri": "[concat(parameters('_artifactsLocation'), '/Api/wiki/wiki.template.json', parameters('_artifactsLocationSasToken'))]",
			"contentVersion": "1.0.0.0"
		},
		"parameters": {
			"url, artifacts location, sas token and apim instance name"
		}
	},
	"dependsOn": [
		"echoTemplate"
	]
}
...
```

When it comes to building support for this scenario into the extractor tool, I'm not so sure. Currently the extractor doesn't really deal with existing files and a project structure, so at least some work is required.

**Comment by SSanjeevi** [comment]
@bremnes - did you have any work around or suggestion for this.

Am using creator tool, there also am facing the same issue.

**Comment by kulpras** [comment]
I am also facing same issue. any solution to avoid **Operation on the API is in progress** error

**Comment by bremnes** [comment]
@SSanjeevi, we ended up doing what I've commented earlier. If you look at my first post, then we have a custom tool wrapped around this resource kit. In that tool we add/manipulate the order of apis in `apis.template.json`. In that file we're adding a "dependsOn" on all the individual APIs so that we deploy them sequentially like [described in my last comment](https://github.com/Azure/azure-api-management-devops-resource-kit/issues/340#issuecomment-600882940). Unfortunately that means the deployment will take a bit longer, but at least it's working 100% of the time.

**Comment by tcukanoot** [comment]
Hi @bremnes, We just hit this same issue. Do you have any more information you could share on the custom tool that wraps the resource kit? We'll likely need to implement something similar, unless there has been further updates on the support ticket.  Thanks!

**Comment by glitchassassin** [comment]
Taking a look at this since we just ran into the same issue.

Sounds to me like this could be fixed in the resource kit with the following steps:

- Move the product/api resources to their own template, which depends upon the product template and all API templates
- Create an arbitrary dependency chain, so each product/api resource depends upon the previous one

This will force all the product/api resources to be deployed in series, after the products and APIs.

I'd like to take a crack at implementing this fix. Any concerns with the approach outlined above?

**Comment by glitchassassin** [comment]
While testing this, it looks like the issue also occurs if the APIs have a common tag resource. The tags are deployed before the APIs, but the APIs can still experience the issue if two APIs have a shared tag in the OpenAPI definitions.

Since the definition is contained in the API resource itself, I don't think there's a resolution for this except to deploy the APIs sequentially, rather than in parallel.

**Comment by tcukanoot** [comment]
Figured I'd post it here, in case it may help. Here's the powershell script we came up with to loop through the api folders (SplitAPIs = true), and add the "dependsOn" for each one...generating the sequential deployment.

`
$templatePath = "..\APIM";
$ApiFolderList = Get-ChildItem -Directory $templatePath | Select-Object -Property Name;
$templatePath += "\master.template.json";
$armTemplate = Get-Content -Path $templatePath -Raw | ConvertFrom-Json;

$previousFolder = "First";

$index = 7;

$ApiFolderList = $ApiFolderList | ? { $_.Name -ne "policies" };

for ($i = 0; $i -lt 7; $i++) {
    $armTemplate.resources[$i].properties.parameters.PolicyXMLBaseUrl.value = "[concat(parameters('PolicyXMLBaseUrl'), '/policies')]";
}

foreach ($folder in $ApiFolderList) {
#for ($folderIndex = 0; $folderIndex -lt $ApiFolderList.Count; $folderIndex++) {
    #$output = "'/" + $folder.Name + "/master.template.json'`r`n";
    #$output;

    if ($armTemplate.resources.Length -lt $index + 1) {
        #append new row here
        $tempRow = $armTemplate.resources[$index - 1];
        $armTemplate.resources += ConvertFrom-Json -InputObject (ConvertTo-Json -Depth 100 -InputObject $tempRow)
    }

    #update templateLink in resources/properties
    $armTemplate.resources[$index].properties.templateLink.uri = "[concat(parameters('LinkedTemplatesBaseUrl'), '/$($folder.Name)/master.template.json')]"

    #add LinkedTemplatesBaseUrl in resources/propertes/parameters
    if ($previousFolder.Equals("First")) {
        #$tempLinkedTemplatesBaseUrl = $armTemplate.resources[$index].properties.parameters.PolicyXMLBaseUrl;
        #$tempLinkedTemplatesBaseUrl.value = "[concat(parameters('LinkedTemplatesBaseUrl'), '/$($folder.Name)')]";

        $temp = "
            {
                ""value"":""[concat(parameters('LinkedTemplatesBaseUrl'), '/$($folder.Name)')]""
            }
        ";

        $tempLinkedTemplatesBaseUrl = ConvertFrom-Json -InputObject $temp;

        $armTemplate.resources[$index].properties.parameters | Add-Member -MemberType NoteProperty -Name "LinkedTemplatesBaseUrl" -Value $tempLinkedTemplatesBaseUrl;
    } else {
        $armTemplate.resources[$index].properties.parameters.LinkedTemplatesBaseUrl.value = "[concat(parameters('LinkedTemplatesBaseUrl'), '/$($folder.Name)')]";
    }

    #modify PolicyXMLBaseUrl in resources/properties/parameters
    $armTemplate.resources[$index].properties.parameters.PolicyXMLBaseUrl.value = "[concat(parameters('PolicyXMLBaseUrl'), '/$($folder.Name)/policies')]";

    #update name in resources
    $tempName = $folder.Name + "Template";
    $armTemplate.resources[$index].Name = $tempName

    #modify dependsOn if not first api record
    if (!$previousFolder.Equals("First")) {
        $tempString = $previousFolder + "Template"
        $temp = @($tempString);
        $armTemplate.resources[$index].dependsOn = $temp;
    }

    $previousFolder = $folder.Name;
    $index++;
}

$armTemplate | ConvertTo-Json -Depth 100 | foreach { [System.Text.RegularExpressions.Regex]::Unescape($_) } | Set-Content -Path $templatePath
`

**Comment by adrianhall** [comment]
As part of our clean-up on old issues, we are closing this issue.  If the problem persists with the 1.0.0-beta releases, please open a new issue and provide all the required information.
